[compiler] Instruction reordering

josephsavona opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #29883
* #29882
* __->__ #29863
* #29881

Adds a pass just after DCE to reorder safely reorderable instructions (jsx, primitives, globals) closer to where they are used, to allow other optimization passes to be more effective. Notably, the reordering allows scope merging to be more effective, since that pass relies on two scopes not having intervening instructions — in many cases we can now reorder such instructions out of the way and unlock merging, as demonstrated in the changed fixtures.

The algorithm itself is described in the docblock.

note: This is a cleaned up version of #29579 that is ready for review.

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/align-scopes-within-nested-valueblock-in-array.expect.md:
```diff
@@ -52,11 +52,12 @@ import { Stringify, identity, makeArray, mutate } from "shared-runtime";
   let t1;
-  if ($[0] !== cond1 || $[1] !== cond2 || $[2] !== arr) {
+  if ($[0] !== cond1 || $[1] !== cond2) {
+    const arr = makeArray({ a: 2 }, 2, []);
```
This should have been part of the memoization block, since `arr.map(mutate)` could mutate (we don't know that `arr` is an Array and can't rely on `map` being immutable.

The new code is correct but i want to double-check why the previous output was incorrect (since the comment suggests it was fixed, and i recall that we did fix this case recently)

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/allow-global-mutation-in-effect-indirect-usecallback.expect.md:
```diff
@@ -67,19 +67,16 @@ function Component() {
-  let t4;
-  if ($[4] === Symbol.for("react.memo_cache_sentinel")) {
     t4 = [someGlobal];
+    $[3] = t3;
```
this case is strictly better, likely from an intermediate instruction getting reordered

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/allow-global-mutation-in-effect-indirect.expect.md:
```diff
@@ -39,60 +39,51 @@ import { useEffect, useState } from "react";
+  let t1;
   if ($[0] === Symbol.for("react.memo_cache_sentinel")) {
-    t0 = () => {
+    const setGlobal = () => {
```
nice improvements, grouping more scopes together

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/dont-merge-if-dep-is-inner-declaration-of-previous-scope.expect.md:
```diff
@@ -99,41 +99,41 @@ function Component(t0) {
     $[15] = t3;
   } else {
     t3 = $[15];
   }
   let t4;
-  if ($[16] !== a || $[17] !== b) {
-    t4 = [a, b];
-    $[16] = a;
-    $[17] = b;
+  if ($[16] !== t2 || $[17] !== x) {
+    t4 = <ValidateMemoization inputs={t2} output={x} />;
+    $[16] = t2;
+    $[17] = x;
     $[18] = t4;
   } else {
     t4 = $[18];
   }
   let t5;
-  if ($[19] !== t4 || $[20] !== z) {
-    t5 = <ValidateMemoization inputs={t4} output={z} />;
-    $[19] = t4;
+  if ($[19] !== t3 || $[20] !== z) {
+    t5 = <ValidateMemoization inputs={t3} output={z} />;
+    $[19] = t3;
```
this example would be more optimal if we could reorder the array literals too, and move them closer to the `<ValidateMemoization>`, but that gets complicated if the arrays could be mutated

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/destructuring-mixed-scope-declarations-and-locals.expect.md:
```diff
@@ -30,47 +30,35 @@ function Component(props) {
-    $[5] = media;
-    $[6] = onClick;
-    $[7] = t0;
+    $[0] = post;
```
the merging here is maybe not ideal - now the `<Media>` element always invalidates whenever `post` does, rather than only if the specific properties change

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/destructuring-mixed-scope-and-local-variables-with-default.expect.md:
```diff
@@ -65,67 +65,52 @@ function Component(props) {
-    $[12] = allUrls;
-    $[13] = onClick;
-    $[14] = t1;
+    $[1] = post;
```
the scope merging here is also not ideal, the `<Stringify>` element only strictly has to invalidate if these properties change, not for all changes to `post`

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/inner-memo-value-not-promoted-to-outer-scope-dynamic.expect.md:
```diff
@@ -24,69 +24,42 @@ function Component(props) {
   if ($[0] !== item) {
     const count = new MaybeMutable(item);
 
-    T1 = View;
-    T0 = View;
-    if ($[5] === Symbol.for("react.memo_cache_sentinel")) {
-      t1 = <span>Text</span>;
-      $[5] = t1;
-    } else {
-      t1 = $[5];
-    }
```
all of this reorders out of the scope, with a huge drop in number of memo slots needed

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/original-reactive-scopes-fork/invalid-align-scopes-within-nested-valueblock-in-array.expect.md:
```diff
@@ -52,11 +52,12 @@ import { Stringify, identity, makeArray, mutate } from "shared-runtime";
   let t1;
-  if ($[0] !== cond1 || $[1] !== cond2 || $[2] !== arr) {
+  if ($[0] !== cond1 || $[1] !== cond2) {
+    const arr = makeArray({ a: 2 }, 2, []);
```
the previous output was incorrect, but this has `@enableReactiveScopesInHIR:false`

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/repro-no-value-for-temporary.expect.md:
```diff
@@ -17,34 +17,30 @@ function Component(listItem, thread) {
     const isFoo = isFooThread(thread.threadType);
-    t1 = useBar;
-    t2 = listItem;
+    t1 = listItem;
```
this is presumably a LoadLocal that we can't reorder (yet)

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/repro-instruction-part-of-already-closed-scope.expect.md:
```diff
@@ -47,62 +46,60 @@ function Component(t0) {
     const b = identity(data, index);
     const c = identity(data, index);
 
-    const t4 = identity(b);
-    if ($[6] !== t4) {
-      t2 = <Stringify value={t4} />;
-      $[6] = t4;
-      $[7] = t2;
-    } else {
-      t2 = $[7];
-    }
-    const t5 = identity(a);
-    if ($[8] !== t5) {
-      t3 = <Stringify value={t5} />;
-      $[8] = t5;
-      $[9] = t3;
-    } else {
-      t3 = $[9];
-    }
```
This is fun: we avoid some of the complications from MergeOverlappingScopes and JSX, now that the JSX instructions shift past later mutating instructions

mvitousek commented on compiler/packages/babel-plugin-react-compiler/src/Optimization/InstructionReordering.ts:
```diff
@@ -0,0 +1,370 @@
+    case "LoadGlobal":
+    case "Primitive":
+    case "TemplateLiteral": {
+      return Reorderability.Reorderable;
```
BinaryExpression?

josephsavona replied:
Oh good call. UnaryExpression too

mvitousek replied:
Was uncertain about this but makes sense--opened #29880 to resolve my uncertainty

josephsavona replied:
nice! Yeah, we had already split out PropertyDelete but i thought that `throw` as an expression wasn't part of the spec yet? but in any case, good to have a todo for it in advance!

mvitousek reviewed (approved):
Extremely cool! Left a couple suggestions and places where I wanted to check my understanding of the algorithm.

josephsavona commented on compiler/packages/babel-plugin-react-compiler/src/Optimization/InstructionReordering.ts:
```diff
@@ -0,0 +1,370 @@
+  const shared: Nodes = new Map();
+  for (const [, block] of fn.body.blocks) {
+    reorderBlock(fn.env, block, shared);
+  }
```
good point, i added a comment and an invariant for this

josephsavona commented:
I updated to:
* address comments
* split out the changes to MergeScopesThatInvalidateTogether into a separate diff, since that is the bulk of the win
* put reordering behind a feature flag
